### PR TITLE
ARROW-13689: [C#][Integration] Initial commit of C# Integration tests

### PR DESCRIPTION
**Pull request by eerhardt**
cc @pitrou - if you want to take a first look. I'm not sure how to get the tests running locally yet, but I've run the integration app a few times against the 'primitive.json' file to test it out.

### CODE REVIEW

**Comment by pitrou** [comment]
> cc @pitrou - if you want to take a first look. I'm not sure how to get the tests running locally yet, but I've run the integration app a few times against the 'primitive.json' file to test it out.

Sorry for the delay. If you're talking about the integration tests as they are run in CI, you can run them using `archery docker run conda-integration`. See https://arrow.apache.org/docs/developers/archery.html for installing Archery.

**Comment by pitrou** [comment]
Ah, you'll need to apply the following patch:
```diff
diff --git a/ci/scripts/integration_arrow.sh b/ci/scripts/integration_arrow.sh
index 453561d31..30cbb2d63 100755
--- a/ci/scripts/integration_arrow.sh
+++ b/ci/scripts/integration_arrow.sh
@@ -28,6 +28,7 @@ pip install -e $arrow_dir/dev/archery
 archery integration \
     --run-flight \
     --with-cpp=1 \
+    --with-csharp=1 \
     --with-java=1 \
     --with-js=1 \
     --with-go=1 \
```

**Comment by pitrou** [comment]
You may also temporarily tweak `ci/scripts/integration_arrow.sh` to disable some implementations, if you want to iterate faster.
(since C++ is the most featureful implementation, you could for example disable Java, Javascript, Go and Rust)

**Comment by eerhardt** [comment]
Thanks for the help @pitrou! I'll keep working on this PR.

**Comment by eerhardt** [comment]
I finally completed the initial work here to test the C# library in the integration tests. I'm not certain whom I should tag for a review of the infra changes I made here. @pitrou @kszucs - can you suggest the appropriate people?

Also tagging recent contributors to the C# library, in case they have any feedback:

@pgovind @suhsteve @chutchinson @Ulimo @mr-smidge @zgramana @HashidaTKS @nhustler @0x0L @llerrowa

File: dev/archery/archery/cli.py [hunk 1]
```diff
@@ -692,6 +692,8 @@ def _set_default(opt, default):
               help="Seed for PRNG when generating test data")
 @click.option('--with-cpp', type=bool, default=False,
               help='Include C++ in integration tests')
+@click.option('--with-csharp', type=bool, default=False,
+              help='Include CSharp in integration tests')
```
Comment (pitrou) [inline dev/archery/archery/cli.py:696]: Should it be spelled "C#" in this message?

File: dev/archery/archery/integration/datagen.py [hunk 2]
```diff
@@ -1557,82 +1557,102 @@ def _temp_path():
...
         generate_null_trivial_case([0, 0])
+        .skip_category('csharp')
         .skip_category('Go')    # TODO(ARROW-7901)
         .skip_category('JS'),   # TODO(ARROW-7900)
 
         generate_decimal128_case()
         .skip_category('Rust'),
 
         generate_decimal256_case()
+        .skip_category('csharp')
```
Comment (pitrou) [inline dev/archery/archery/integration/datagen.py:1579]: According to https://github.com/apache/arrow/blob/master/docs/source/status.rst#data-types, Decimal256 should be supported?

File: dev/archery/archery/integration/tester_csharp.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+
+class CSharpTester(Tester):
+    PRODUCER = True
+    CONSUMER = True
+
+    EXE_PATH = os.path.join(
+        ARROW_ROOT_DEFAULT, 'csharp/artifacts/Apache.Arrow.IntegrationTest/Debug/netcoreapp3.1/Apache.Arrow.IntegrationTest')
+
+    name = 'csharp'
```
Comment (pitrou) [inline dev/archery/archery/integration/tester_csharp.py:31]: Wouldn't "C#" be a better name for display?

**Review by pitrou — COMMENTED** [review]
The infrastructure changes look fine to me, just a couple questions.

File: dev/archery/archery/cli.py (hunk 1 above)
Comment (eerhardt) [reply dev/archery/archery/cli.py:696]: Yes, good call.

File: dev/archery/archery/integration/datagen.py (hunk 2 above)
Comment (eerhardt) [reply dev/archery/archery/integration/datagen.py:1579]: It is supported. However, I needed to cut things off somewhere in order to get the initial implementation out. If it is easy to implement now that I have everything working, I'll implement it.

File: dev/archery/archery/integration/tester_csharp.py (hunk 3 above)
Comment (eerhardt) [reply dev/archery/archery/integration/tester_csharp.py:31]: Yes, good call.

File: dev/archery/archery/integration/datagen.py (hunk 2 above)
Comment (eerhardt) [reply dev/archery/archery/integration/datagen.py]: It was pretty simple to implement this now, so I did.

**Comment by eerhardt** [comment]
To move forward, I'll merge this either late today or tomorrow. Let me know if you wanted to review it and need more time.

**Comment by pitrou** [comment]
I'll try to take a second look quickly.

**Comment by pitrou** [comment]
No objection from me. Thanks _a lot_ for working on this!

**Pull request closed without merging**